File: packages/core/src/types/document.ts
import { z } from "zod";

export interface Citation { sourceId: string; locator: string }
export interface Span { text: string; citation?: Citation }
export type Block =
  | { type: "paragraph"; spans: Span[] }
  | { type: "table"; columns: string[]; rows: { cells: Span[][] }[] };
export interface DocSection { key: string; heading: string; blocks: Block[] }
export interface RunDocument { title: string; eyebrow: string; dateline: string; sections: DocSection[] }

export function blocksToPlainText(blocks: Block[]): string {
  return blocks.map((b) =>
    b.type === "paragraph"
      ? b.spans.map((s) => s.text).join("")
      : b.rows.map((r) => r.cells.map((c) => c.map((s) => s.text).join("")).join(" · ")).join("\n")
  ).join("\n\n");
}

export function countWords(blocks: Block[]): number {
  return blocksToPlainText(blocks).split(/\s+/).filter(Boolean).length;
}

// ---- zod schemas (v1.5): validation for the shapes above ---------------------
const KEY_RE = /^[a-z][a-z0-9_]*$/;

export const CitationSchema = z.object({ sourceId: z.string().min(1), locator: z.string().min(1) }).strict();
export const SpanSchema = z.object({ text: z.string(), citation: CitationSchema.optional() }).strict();
export const BlockSchema = z.discriminatedUnion("type", [
  z.object({ type: z.literal("paragraph"), spans: z.array(SpanSchema) }).strict(),
  z.object({
    type: z.literal("table"),
    columns: z.array(z.string()),
    rows: z.array(z.object({ cells: z.array(z.array(SpanSchema)) }).strict()),
  }).strict(),
]);
export const DocSectionSchema = z.object({
  key: z.string().regex(KEY_RE), heading: z.string().min(1), blocks: z.array(BlockSchema),
}).strict();
export const RunDocumentSchema = z.object({
  title: z.string(), eyebrow: z.string(), dateline: z.string(), sections: z.array(DocSectionSchema),
}).strict().superRefine((doc, ctx) => {
  const seen = new Set<string>();
  for (const s of doc.sections) {
    if (seen.has(s.key)) ctx.addIssue({ code: z.ZodIssueCode.custom, message: `duplicate section key: ${s.key}` });
    seen.add(s.key);
  }
});

File: backend/runoff_api/core/types/document.py
"""Port of packages/core/src/types/document.ts (v1.5 doc shapes + zod schemas)."""

import re
from typing import Annotated, Literal

from pydantic import Field, model_validator

from .base import CamelModel

KEY_RE = r"^[a-z][a-z0-9_]*$"


class Citation(CamelModel):
    source_id: str = Field(min_length=1)
    locator: str = Field(min_length=1)


class Span(CamelModel):
    text: str
    citation: Citation | None = None


class ParagraphBlock(CamelModel):
    type: Literal["paragraph"]
    spans: list[Span]


class TableRowModel(CamelModel):
    cells: list[list[Span]]


class TableBlock(CamelModel):
    type: Literal["table"]
    columns: list[str]
    rows: list[TableRowModel]


Block = Annotated[ParagraphBlock | TableBlock, Field(discriminator="type")]


class DocSection(CamelModel):
    key: str = Field(pattern=KEY_RE)
    heading: str = Field(min_length=1)
    blocks: list[Block]


class RunDocument(CamelModel):
    title: str
    eyebrow: str
    dateline: str
    sections: list[DocSection]

    @model_validator(mode="after")
    def _unique_section_keys(self) -> "RunDocument":
        seen: set[str] = set()
        for s in self.sections:
            if s.key in seen:
                raise ValueError(f"duplicate section key: {s.key}")
            seen.add(s.key)
        return self


def blocks_to_plain_text(blocks: list[dict]) -> str:
    """Port of blocksToPlainText — operates on plain dicts (runtime shape)."""
    out = []
    for b in blocks:
        if b["type"] == "paragraph":
            out.append("".join(s["text"] for s in b["spans"]))
        else:
            out.append("\n".join(
                " · ".join("".join(s["text"] for s in cell) for cell in row["cells"])
                for row in b["rows"]
            ))
    return "\n\n".join(out)


def count_words(blocks: list[dict]) -> int:
    """Port of countWords."""
    return len([w for w in re.split(r"\s+", blocks_to_plain_text(blocks)) if w])
